### PR TITLE
system implicit routing + routing path type

### PR DESCRIPTION
**Pull request by adisos**
issue #496

### CODE REVIEW

File: pkg/ibmvpc/routing_paths.go
Comment (zivnevo) [inline pkg/ibmvpc/routing_paths.go:1]: Shouldn't this reside in `vpcmodel`? Looks general enough

File: pkg/ibmvpc/routing_paths.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	case e.nextHop != nil:
+		return e.nextHop.string()
+	}
+	return ""
+}
+
+func (e *endpoint) equal(otherEndpoint *endpoint) bool {
+	switch {
+	case e.nextHop != nil:
+		if !otherEndpoint.nextHop.equal(e.nextHop) {
```
Comment (zivnevo) [inline pkg/ibmvpc/routing_paths.go:77]: `otherEndpoint.nextHop` might be nil

File: pkg/ibmvpc/routing_paths.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+			return false
+		}
+	case e.vpcResource != nil:
+		if otherEndpoint.vpcResource == nil || otherEndpoint.vpcResource.UID() != e.vpcResource.UID() {
+			return false
+		}
+	case
+		e.ipBlock != nil:
```
Comment (zivnevo) [inline pkg/ibmvpc/routing_paths.go:85]:
```suggestion
	case e.ipBlock != nil:
```

File: pkg/ibmvpc/implicit_routing.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+/*
+You cannot configure a routing table to use the system-implicit routing table; it is populated automatically.
+The system-implicit routing table is used when no matching route is found in the custom routing table that is
+associated with the subnet where the traffic is egressing.
+If no match is found, the packet is dropped.
+
+This behavior can be avoided with a custom routing table default route with an action of drop.
+*/
+type systemImplicitRT struct {
+	vpc    *VPC
```
Comment (zivnevo) [inline pkg/ibmvpc/implicit_routing.go:34]:
```suggestion
	vpc    *VPC // parent VPC
```

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/ibmvpc/implicit_routing.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	config *config
+	// TODO: should be per zone in vpc
+}
+
+type config struct {
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/implicit_routing.go]: right, changed

File: pkg/ibmvpc/routing_paths.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+		n.origDest.Equal(other.origDest)
+	// TODO: add comparison of rt ?
+}
+
+func ipBlockString(ipb *ipblock.IPBlock) string {
+	if ipAddress := ipb.ToIPAddressString(); ipAddress != "" {
+		return ipAddress
+	}
+	return ipb.ToCidrListString()
+}
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/routing_paths.go]: opened PR there: https://github.com/np-guard/models/pull/31

File: pkg/ibmvpc/routing_paths.go (hunk 1 above)
Comment (adisos) [reply pkg/ibmvpc/routing_paths.go]: right, fixed

File: pkg/ibmvpc/routing_paths.go (hunk 2 above)
Comment (adisos) [reply pkg/ibmvpc/routing_paths.go]: done

File: pkg/ibmvpc/implicit_routing.go (hunk 3 above)
Comment (adisos) [reply pkg/ibmvpc/implicit_routing.go]: done

File: pkg/ibmvpc/routing_paths.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/routing_paths.go]: done

**Review by zivnevo — APPROVED** [review]

**Pull request merged**